File: vueapp/src/models/windowProfile.js
import CrudModel from './crudModel.js';
import Brand from './brand.js';
import Image from './image.js';
import WindowProfileColor from './windowProfileColor.js';

import { reactive } from 'vue'

import { isEmpty } from 'lodash-es';

class WindowProfile extends CrudModel {
	constructor(json = {}) {
		super(json);
		this.name = json.name || null;
		this.description = json.description || null;
		this.brandId = json.brandId || null;
		this.brand = json.brand ? new Brand(json.brand) : null;
		this.imageId = json.imageId || null;
		this.image = json.image ? new Image(json.image) : (!isEmpty(this.imageId) ? new Image({ id: this.imageId }) : null);
		this.thermalInsulation = json.thermalInsulation || 0.0;
		this.soundInsulation = json.soundInsulation || 0;
		this.antiTheftResistance = json.antiTheftResistance || null;
		this.constructionDepth = json.constructionDepth || 0;
		this.frameChamberCount = json.frameChamberCount || 0;
		this.airPermeability = json.airPermeability || null;
		this.waterTightness = json.waterTightness || null;
		this.windResistance = json.windResistance || null;
		this.maxGlassThickness = json.maxGlassThickness || 0.0;
		this.colors = json.colors ? json.colors.map((color) => new WindowProfileColor(color)) : null;

		this.sideHungOpening = json.sideHungOpening || false;
		this.tiltAndTurnOpening = json.tiltAndTurnOpening || false;
		this.tiltOnlyOpening = json.tiltOnlyOpening || false;
		this.tiltAndParallelOpening = json.tiltAndParallelOpening || false;

		this.maxLeafSizeWidth = json.maxLeafSizeWidth || 0;
		this.maxLeafSizeHeight = json.maxLeafSizeHeight || 0;
	}

	hydrate(data)
    {
        super.hydrate(data)

        if(!this.image && !isEmpty(this.imageId))
            reactive(this).image = reactive(new Image({ id: this.imageId }))
    }

	toFormData() {
		const formData = super.toFormData();

		formData.append('name', this.getSafeStringValue(this.name));
		formData.append('description', this.getSafeStringValue(this.description));
		formData.append('brandId', this.getSafeKeyValue(this.brandId));
		formData.append('imageId', this.getSafeKeyValue(this.imageId));
        if(this.image?.isDirty === true) // Só se houver alteração é que vamos submeter a nova imagem
            this.image?.toOtherFormData(formData, 'image');
		formData.append('thermalInsulation', this.thermalInsulation);
		formData.append('soundInsulation', this.soundInsulation);
		formData.append('antiTheftResistance', this.getSafeStringValue(this.antiTheftResistance));
		formData.append('constructionDepth', this.constructionDepth);
		formData.append('frameChamberCount', this.frameChamberCount);
		formData.append('airPermeability', this.getSafeStringValue(this.airPermeability));
		formData.append('waterTightness', this.getSafeStringValue(this.waterTightness));
		formData.append('windResistance', this.getSafeStringValue(this.windResistance));
		formData.append('maxGlassThickness', this.maxGlassThickness);

		formData.append('sideHungOpening', this.sideHungOpening);
		formData.append('tiltAndTurnOpening', this.tiltAndTurnOpening);
		formData.append('tiltOnlyOpening', this.tiltOnlyOpening);
		formData.append('tiltAndParallelOpening', this.tiltAndParallelOpening);

		formData.append('maxLeafSizeWidth', this.maxLeafSizeWidth);
		formData.append('maxLeafSizeHeight', this.maxLeafSizeHeight);

		return formData;
	}

	toSrvData()
    {
        return this.toFormData()
    }

	setImage({file, fileName, fileType})
    {
        this.image = new Image({
            id: this.imageId,
            file,
            fileName,
            fileType,
            isDirty: true
        })
    }
}

export default WindowProfile;

File: webapi/Models/WindowProfile.cs
using Microsoft.EntityFrameworkCore;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace webapi.Models
{
    /// <summary>
    /// Representa um perfil de janela.
    /// </summary>
    public class WindowProfile : CrudModel
    {
        /// <summary>
        /// O nome do perfil da janela.
        /// </summary>
        [Required, MaxLength(255)]
        public string? Name { get; set; }

        /// <summary>
        /// Descrição do perfil da janela.
        /// </summary>
        [MaxLength(10000)]
        public string? Description { get; set; }

        [ForeignKey("Brand"), Required(ErrorMessage = "A marca do perfil da janela é obrigatório. Por favor, preencha o campo.")]
        public Guid? BrandId { get; set; }

        /// <summary>
        /// Marca
        /// </summary>
        public virtual Brand? Brand { get; set; }

        public Guid? ImageId { get; set; }

        /// <summary>
        /// Imagem do perfil
        /// </summary>
        [ForeignKey(nameof(ImageId))]
        public Image? Image { get; set; }

        /// <summary>
        /// Isolamento térmico - Uf (W/m²K)
        /// </summary>
        public double ThermalInsulation { get; set; } = 0;

        /// <summary>
        /// Isolamento acústico - Rw (dB)
        /// </summary>
        public int SoundInsulation { get; set; } = 0;

        /// <summary>
        /// Resistência antirroubo
        /// </summary>
        public string? AntiTheftResistance { get; set; }

        /// <summary>
        /// Profundidade construtiva (mm)
        /// </summary>
        public int ConstructionDepth { get; set; } = 0;

        /// <summary>
        /// Número de câmaras
        /// </summary>
        public int FrameChamberCount { get; set; } = 0;

        /// <summary>
        /// Permeabilidade ao ar
        /// </summary>
        public string? AirPermeability { get; set; }

        /// <summary>
        /// Estanqueidade à água
        /// </summary>
        public string? WaterTightness { get; set; }

        /// <summary>
        /// Resistência ao vento
        /// </summary>
        public string? WindResistance { get; set; }

        /// <summary>
        /// Espessura máxima do vidro suportada (mm)
        /// </summary>
        public int MaxGlassThickness { get; set; } = 0;

        /// <summary>
        /// Suporta abertura Batente
        /// </summary>
        public bool SideHungOpening { get; set; }
        /// <summary>
        /// Suporta abertura Oscilobatente
        /// </summary>
        public bool TiltAndTurnOpening { get; set; }
        /// <summary>
        /// Suporta abertura Basculante
        /// </summary>
        public bool TiltOnlyOpening { get; set; }
        /// <summary>
        /// Suporta abertura Oscilo-paralelo
        /// </summary>
        public bool TiltAndParallelOpening { get; set; }

        /// <summary>
        /// Dimensões máx. folha - Largura (mm)
        /// </summary>
        public int MaxLeafSizeWidth { get; set; } = 0;
        /// <summary>
        /// Dimensões máx. folha - Altura (mm)
        /// </summary>
        public int MaxLeafSizeHeight { get; set; } = 0;

        /// <summary>
        /// Lista de cores compatíveis.
        /// </summary>
        [DeleteBehavior(DeleteBehavior.Cascade)]
        public virtual IList<WindowProfileColor>? Colors { get; set; }

        public WindowProfile() : base()
        {
        }
    }
}
